### ISSUE TITLE
Memory leak when Ceph connect args is wrong

### ISSUE THREAD

**Issue by iawia002** [ceph/go-ceph#109]
Here is my test code:

```go
func getConnection(monitors, user, key string) (*rados.Conn, error) {
	conn, err := rados.NewConnWithUser(user)
	if err != nil {
		return nil, err
	}
	args := []string{"--client_mount_timeout", "15", "-m", monitors, "--key", key}
	err = conn.ParseCmdLineArgs(args)
	if err != nil {
		return nil, err
	}
	err = conn.Connect()
	if err != nil {
		return nil, err
	}
	return conn, nil
}
```

Memory leak occurs when this function is executed frequently with the wrong `args` (eg: wrong key). But when the args are correct, there is no memory leak.

So my guess is that there was no cleanup operation after called `C.rados_connect` with the wrong `args`. I try to change the `Shutdown` function(remove the `ensure_connected` check) and execute it every time `Connect` failed, and the problem was solved.

```diff
 func (c *Conn) Shutdown() {
-       if err := c.ensure_connected(); err != nil {
-               return
-       }
        C.rados_shutdown(c.cluster)
 }
```

I don't know if this is a bug or there is something wrong with my usage.

**Comment by Madhu-1** [comment]
CC @nixpanic @phlogistonjohn

**Comment by nixpanic** [comment]
I think the problem lays within `ParseCmdLineArgs()`. The call inside `librados` configures a rados client with the arguments, but if that fails, the (partially) configured arguments are not freed. I do not immediately see how this can be done within `librados`. But in any case, we should probably call `rados_shutdown()` on the *Conn when it goes out of scope so that the connections get closed automatically (if connected) and other resources are freed.

I'll have a go at getting that done.

**Comment by nixpanic** [comment]
The following test can be used to check for the memory leak:

```go
// issue-109.go: analyze memory leak when rados.Conn.Connect() fails.
//
// build with: go build issue-109.go
// test with: valgrind ./issue-109

package main

import (
	"fmt"
	"runtime"
	"time"

	"github.com/ceph/go-ceph/rados"
)

func main() {
	for i := 1; i <= 100; i++ {
		c, err := getConnection("127.0.0.1", "nobody", "secret")
		if err != nil {
			fmt.Printf("getConnection failed: %v\n", err)
		} else {
			c.Shutdown()
		}
	}

	// force running the garbage collector
	runtime.GC()
	time.Sleep(time.Second)
}

func getConnection(monitors, user, key string) (*rados.Conn, error) {
	conn, err := rados.NewConnWithUser(user)
	if err != nil {
		return nil, err
	}
	args := []string{"--client_mount_timeout", "15", "-m", monitors, "--key", key}
	err = conn.ParseCmdLineArgs(args)
	if err != nil {
		return nil, fmt.Errorf("ParseCmdLineArgs: %v", err)
	}
	err = conn.Connect()
	if err != nil {
		return nil, fmt.Errorf("Connect: %v", err)
	}
	return conn, nil
}
```

Original problem identified with valgrind:
```
==229515== 
==229515== HEAP SUMMARY:
==229515==     in use at exit: 1,200,029,188 bytes in 565,672 blocks
==229515==   total heap usage: 1,410,377 allocs, 844,705 frees, 1,346,477,310 bytes allocated
==229515== 
==229515== LEAK SUMMARY:
==229515==    definitely lost: 4,000 bytes in 100 blocks
==229515==    indirectly lost: 7,200 bytes in 300 blocks
==229515==      possibly lost: 137,188 bytes in 326 blocks
==229515==    still reachable: 1,199,880,800 bytes in 564,946 blocks
==229515==         suppressed: 0 bytes in 0 blocks
```

With fix in PR #110:
```
==230686== HEAP SUMMARY:
==230686==     in use at exit: 552,593,158 bytes in 262,749 blocks
==230686==   total heap usage: 1,414,354 allocs, 1,151,605 frees, 1,346,655,182 bytes allocated
==230686== 
==230686== LEAK SUMMARY:
==230686==    definitely lost: 4,000 bytes in 100 blocks
==230686==    indirectly lost: 7,200 bytes in 300 blocks
==230686==      possibly lost: 64,660 bytes in 164 blocks
==230686==    still reachable: 552,517,298 bytes in 262,185 blocks
==230686==         suppressed: 0 bytes in 0 blocks
```

There are still some leaks or other un-free'd resources, but the calling of `rados_shutdown()` releases 647,363,502 bytes over 100 iterations in the test.
